lightning/src/ln/outbound_payment.rs

```diff
@@ -597,21 +623,21 @@ impl OutboundPayments {
 			onion_session_privs.push(entropy_source.get_secure_random_bytes());
 		}
 
-		let (total_msat, payment_hash, payment_secret) = {
+		let (total_msat, payment_hash, payment_secret, keysend_preimage) = {
 			let mut outbounds = self.pending_outbound_payments.lock().unwrap();
 			match outbounds.get_mut(&payment_id) {
 				Some(payment) => {
 					let res = match payment {
 						PendingOutboundPayment::Retryable {
-							total_msat, payment_hash, payment_secret, pending_amt_msat, ..
+							total_msat, payment_hash, keysend_preimage, payment_secret, pending_amt_msat, ..
 						} => {
 							let retry_amt_msat: u64 = route.paths.iter().map(|path| path.last().unwrap().fee_msat).sum();
 							if retry_amt_msat + *pending_amt_msat > *total_msat * (100 + RETRY_OVERFLOW_PERCENTAGE) / 100 {
 								return Err(PaymentSendFailure::ParameterError(APIError::APIMisuseError {
 									err: format!("retry_amt_msat of {} will put pending_amt_msat (currently: {}) more than 10% over total_payment_amt_msat of {}", retry_amt_msat, pending_amt_msat, total_msat).to_string()
 								}))
 							}
-							(*total_msat, *payment_hash, *payment_secret)
+							(*total_msat, *payment_hash, *payment_secret, *keysend_preimage)
```

Comment:
Possible follow-up opportunity to improve code cleanliness here. The first pattern is the only one that doesn't return early. So some of the indentations can be reduced by getting rid of the `res` local variable (and thus moving up the `match`) and also the outer `(total_msat, payment_hash, payment_secret, keysend_preimage)` variables, though you'd have to drop the lock before calling `pay_route_internal`.